### PR TITLE
2016 day 8 

### PR DESCRIPTION
**Pull request by lifehug**
Created a library for 2016 day 8 which creates a Matrix along with supporting functions like inserting, rotating, building and more.  

- Why
To finish the Advent of Code - of course!

## PR - Merge Checklist
- [ X] CR'd
- [ X] Workflow: Trello card is in "Code Review" column
- [ X] README updated or N/A
- [ X] Benchmarks run and compared with master or N/A

### CODE REVIEW

**Comment by lifehug** [comment]
[2016-day-8-part-1](https://trello.com/c/EQHVHka2/113-2016-day-8-part-1)

**Comment by lifehug** [comment]
[2016-day-8-part-2](https://trello.com/c/n0AuUUW9/138-2016-day-8-part-2)

File: src/year2016/day_8.rs [hunk 1]
```diff
@@ -0,0 +1,109 @@
+use regex::Regex;
+use std::fmt;
+
+#[derive(Eq, PartialEq, Hash, Debug, Clone, Default)]
+pub struct Matrix {
+    m: Vec<std::vec::Vec<u32>>,
```
Comment (kanekv) [inline src/year2016/day_8.rs]: why `std::vec::Vec` is there?

File: src/year2016/day_8.rs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+                }
+            }
+        }
+
+        Matrix { m: self.m.clone(), c: self.c, r: self.r}
+    }
+
+    pub fn rotate_column(&mut self, c: usize, n: usize) -> Matrix {
+        // rotates a column of the matrix by a given number
+        let mut column = self.m.iter().fold(Vec::new(), |mut v, r| {
```
Comment (kanekv) [inline src/year2016/day_8.rs]: the way you did is fine, FYI there is another way to do it with `map(...).collect()`

File: src/year2016/day_8.rs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    pub fn rotate_row(&mut self, r: usize, n: usize) -> Matrix {
+        // rotates a row by the given number
+        self.m[r].rotate_right(n);
+         Matrix {m: self.m.clone(), c: self.c, r: self.r}
+    }
+
+    pub fn build(vector: Vec<std::vec::Vec<u32>>) -> Matrix {
+        // creates a Matrix from the given nexted vector
```
Comment (kanekv) [inline src/year2016/day_8.rs]: `nexted` -> `nested`

File: src/year2016/day_8.rs [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        let r = vector.len();
+        
+        Matrix {m: vector, c, r}
+    }
+
+    pub fn count(&self) -> u32 {
+        self.m.iter().fold(0, |acc, r| acc + r.iter().sum::<u32>())
+    }
+
+    // gotta clean up this function a little
```
Comment (kanekv) [inline src/year2016/day_8.rs]: I prefer to prefix these comments with `TODO:`

File: src/year2016/day_8.rs (hunk 1 above)
Comment (lifehug) [reply src/year2016/day_8.rs]: ah, copy and paste.

File: src/year2016/day_8.rs (hunk 2 above)
Comment (lifehug) [reply src/year2016/day_8.rs]: that's good to know, that is a little easier to read too

**Pull request merged**